packages/swing-store/src/bundleStore.js

```diff
@@ -62,20 +66,36 @@ export function makeBundleStore(db, ensureTxn, noteExport = () => {}) {
     return `b${Nat(version)}-${hash}`;
   }
 
-  const sqlAddBundle = db.prepare(`
-    INSERT OR REPLACE INTO bundles
-      (bundleID, bundle)
-    VALUES (?, ?)
+  // the PRIMARY KEY constraint requires the bundleID not already
+  // exist
+  const sqlAddBundleRecord = db.prepare(`
+INSERT INTO bundles (bundleID, bundle) VALUES (?, NULL)
   `);
 
-  /**
-   * Store a bundle.  Here the bundle itself is presumed valid.
-   *
-   * @param {string} bundleID
-   * @param {Bundle} bundle
-   */
-  function addBundle(bundleID, bundle) {
+  // this sees both populated and pruned (not-yet-populated) records
+  const sqlHasBundleRecord = db.prepare(`
+    SELECT count(*)
+    FROM bundles
+    WHERE bundleID = ?
+  `);
+  sqlHasBundleRecord.pluck();
+
+  const sqlPopulateBundleRecord = db.prepare(`
+    UPDATE bundles SET bundle = ? WHERE bundleID = ?
+  `);
+
+  function addBundleRecord(bundleID) {
+    ensureTxn();
+    sqlAddBundleRecord.run(bundleID);
+  }
+
+  function populateBundle(bundleID, serialized) {
     ensureTxn();
+    sqlHasBundleRecord.run(bundleID) || Fail`missing ${bundleID}`;
+    sqlPopulateBundleRecord.run(serialized, bundleID);
```

Comment:
SQLite has a named parameter thing, but since the `sqlPopulateBundleRecord` definition is so close to the invocation, I didn't bother to use it. https://www.sqlite.org/lang_expr.html#varparam plus https://github.com/WiseLibs/better-sqlite3/blob/master/docs/api.md#binding-parameters , so something like:
```js
  const sqlPopulateBundleRecord = db.prepare(`
    UPDATE bundles SET bundle = $serialized WHERE bundleID = $bundleID
  `);

  ..

    sqlPopulateBundleRecord.run({ bundleID, serialized });
```

.. actually, that looks kinda nice, I'll switch to it.

Reply:
Oh I wish we had done that everywhere! I hate positional params